src/signed.rs

```diff
@@ -0,0 +1,68 @@
+use crate::{Index, KeyBox};
+use codec::{Decode, Encode};
+
+pub trait Signable {
```

Comment:
How about adding `AsRef<[u8]>` instead of this trait?

Reply:
`AsRef` is supposed to be for cheap conversions and here we want to perform some encoding. Perhaps `Into<[u8]>` could work, but the semantics are a bit different – perhaps we only want a part of the whole object (in our case maybe the hash?) signed, not the whole representation for whatever reason?

Reply:
Hmm...so maybe `sign<S, F: Fn(&[u8]) -> S>(&self, signing: F) -> S` ?

Reply:
Making `bytes_to_sign` return `&[u8]` seems simpler (can we do that?). We could also autoimplement for anything that is `AsRef<[u8]>` if we really want to.

Reply:
It would be way simpler, but we cannot return references :(

Reply:
...how about we return `impl AsRef<[u8]>` then?

Reply:
But then we need to create a temporary object that will implement this trait and the idea (at least mine) was to change this trait in a way that avoids it.

Reply:
I thought `&[u8]` automatically implements `AsRef<[u8]>`, doesn't it? Then we could return `as_ref` directly for types that implement it.

Reply:
It does but still borrow checker won't let you return type `&T`.

Reply:
Is there still a scenario, where we do not want to sign_to_bytes is essentially different from encode()? maybe we should just replace Signable with Encode?

Reply:
Then we are always copying, aren't we? I think in this case it would be nicer to keep it separate, to keep the semantics clear, but I won't insist.

Whatever we end up with I would prefer avoiding the `sign<...` option, it also requires a `verify<...` function, and both of these have to be implemented by the author of the type to be signed and would always have to contain almost identical logic, with the exception of how it acquires the `&[u8]`. This would cause lots of code duplication or worse, actual differences in the logic in the implementations.

Reply:
If we want to keep the possibility of avoiding copying then we can either do [some magic](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=trait%20Signable%3C%27b%2C%20B%3A%20AsRef%3C%5Bu8%5D%3E%3E%20%7B%0A%20%20%20%20%20fn%20bytes_to_sign(%26%27b%20self)%20-%3E%20B%3B%0A%7D%0A%0Astruct%20Wrapper%3C%27a%2C%20T%3A%20AsRef%3C%5Bu8%5D%3E%3E%20%7B%0A%20%20%20%20data%3A%20%26%27a%20T%2C%0A%7D%0A%0Aimpl%3C%27a%2C%20T%3A%20AsRef%3C%5Bu8%5D%3E%3E%20AsRef%3C%5Bu8%5D%3E%20for%20Wrapper%3C%27a%2C%20T%3E%20%7B%0A%20%20%20%20fn%20as_ref(%26self)%20-%3E%20%26%5Bu8%5D%20%7B%0A%20%20%20%20%20%20%20%20self.data.as_ref()%0A%20%20%20%20%7D%0A%7D%0A%0Aimpl%3C%27b%2C%20T%3A%20AsRef%3C%5Bu8%5D%3E%3E%20Signable%3C%27b%2C%20Wrapper%3C%27b%2C%20T%3E%3E%20for%20T%20%7B%0A%20%20%20%20fn%20bytes_to_sign(%26%27b%20self)%20-%3E%20Wrapper%3C%27b%2C%20T%3E%20%7B%0A%20%20%20%20%20%20%20%20Wrapper%7Bdata%3A%20self%7D%0A%20%20%20%20%7D%0A%7D%0A%0Astruct%20Hello%3B%0A%0Aimpl%3C%27b%3E%20Signable%3C%27b%2C%20Vec%3Cu8%3E%3E%20for%20Hello%20%7B%0A%20%20%20%20fn%20bytes_to_sign(%26%27b%20self)%20-%3E%20Vec%3Cu8%3E%20%7B%0A%20%20%20%20%20%20%20%20%22Hello%22.as_bytes().to_vec()%0A%20%20%20%20%7D%0A%7D%0A%0Astruct%20DummyContainer%3C%27a%2C%20T%3A%20Signable%3C%27a%2C%20Vec%3Cu8%3E%3E%3E%20%7B%0A%20%20%20%20hi%3A%20%26%27a%20T%2C%0A%7D%0A%0Afn%20main()%20%7B%0A%20%20%20%20let%20hi%20%3D%20DummyContainer%7Bhi%3A%20%26Hello%7B%7D%7D%3B%0A%20%20%20%20println!(%22%7B%3A%3F%7D%22%2Chi.hi.bytes_to_sign())%3B%0A%20%20%20%20println!(%22%7B%3A%3F%7D%22%2Chi.hi.bytes_to_sign())%3B%0A%7D), or force the implementer to figure out how to return a reference (e.g. by keeping the bytes to sign as part of the struct).

Reply:
Note that the `Signable` trait is not in the public interface of rush, and if I am not wrong, the only implementors of this trait which we plan for now are (full) units and alerts. Even if we only look at the `FullUnit` type, it seems unavoidable to create a temporary object as we need to call the `encode` method on the data of the unit. Unless we have a reason to think that this will be the bottleneck of rush, I think we should go with the simplest solution, which is making the `sign_to_bytes` method return `Vec<u8>`.

Reply:
Unless we sign a hash of unit, which would save us calling `encode` and creating a temporary object (we should change what we sign regardless of what will happen to `Signable` trait). Then, the easiest solution would be to use `AsRef`, probably without even creating a new trait, just requiring it as a bound in `Signed`. I believe that it fits there pretty well.

Reply:
We decided to give `Signable` a single method  `fn hash(&self) -> Self::Hash`and to sign an object means exactly to sign its hash.